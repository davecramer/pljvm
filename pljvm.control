# pljvm extension
comment = 'execute remote functions inside jvm'
default_version = '1.0.0'
module_pathname = '$libdir/pljvm'
relocatable = true