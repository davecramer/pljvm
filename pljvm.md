#PLJVM PostgreSQL procedural language for JAVA

## PL/JVM enables calling arbitrary java methods on objects running in your code.
- came from [PL/J](https://github.com/codehaus/plj) 
- pl/j was written 14 years ago by Laszlo Hornyak
- competed with pl/java and lost the VHS/Betamax discussion
- since then while working at Pivotal I helped write [pl/container](https://github.com/greenplum-db/plcontainer)
- PL/Container - GPDB execution sandboxing using docker containers for Python and R
- all of these provide what is referred to as a "trusted" language in PostgreSQL. 
- trusted languages are prevented from harming the database by either directly accessing the filesystem or reading others data.
- also PostgreSQL allows developers to create trusted language functions without being a superuser.

## Why ?
- well mostly because I could and I had some free time.
- almost all of the work on the server was done and debugged
- netty makes it somewhat easy (after you grok netty).


## Under the covers
- On the server side you create a language called pl/jvm
- languages are nothing more than a call handler written in C which intercepts a call and has information about 
    - return type
    - arguments
    - function definition

- Once control is transferred to the function handler then the fun begins
    - create a call request object with the arguments
    - send the call request to the Java program using a binary protocol which is very similar to PostgreSQL protocol
	- control is transferred to the Java code which 
	    - decodes the request
        - instantiates the function and calls the method defined in the function call
        - encodes the response and sends it back
        - returns control back to the C code in the server
	- currently the code can deal with any base type and arrays of base types

## What problems does this solve?

- synchronous notification of data changes
- alternatives pl/java logical replication.
- what makes it so special it can be used with any jvm based language
- performance
- need help with 
- classpath separation
- library or full fledged application?
- how to pass UDT's where do I get the package name from
- autoboxing ?