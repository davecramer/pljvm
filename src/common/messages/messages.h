/*------------------------------------------------------------------------------
 *
 *
 * Copyright (c) 2016, Pivotal.
 *
 *------------------------------------------------------------------------------
 */
#ifndef PLC_MESSAGES_H
#define PLC_MESSAGES_H

#include "message_types.h"
#include "message_base.h"
#include "message_callreq.h"
#include "message_result.h"
#include "message_sql.h"
#include "message_error.h"
#include "message_log.h"
#include "message_data.h"
#include "message_ping.h"

#endif /* PLC_MESSAGES_H */
