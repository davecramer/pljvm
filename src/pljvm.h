/*------------------------------------------------------------------------------
 *
 *
 * Copyright (c) 2016, Pivotal.
 *
 *------------------------------------------------------------------------------
 */

#ifndef PLC_PLCONTAINER_H
#define PLC_PLCONTAINER_H

#include "fmgr.h"

#define UNUSED __attribute__ (( unused ))

MemoryContext pl_container_caller_context;

/* entrypoint for all pljvm procedures */
Datum pljvm_call_handler(PG_FUNCTION_ARGS);

#endif /* PLC_PLCONTAINER_H */
