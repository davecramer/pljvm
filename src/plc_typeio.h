/*------------------------------------------------------------------------------
 *
 *
 * Copyright (c) 2016, Pivotal.
 *
 *------------------------------------------------------------------------------
 */


#ifndef PLC_TYPEIO_H
#define PLC_TYPEIO_H

#include "postgres.h"
#include "funcapi.h"

#include "common/messages/messages.h"
#include "pljvm.h"

typedef struct plcTypeInfo plcTypeInfo;
typedef char *(*plcDatumOutput)(Datum, plcTypeInfo*);
typedef Datum (*plcDatumInput)(char*, plcTypeInfo*);

struct plcTypeInfo {
    /* PL/Container-specific information */
    plcDatatype     type;
    int             nSubTypes;
    plcTypeInfo    *subTypes;

    /* Custom input and output functions used for most common data types that
     * allow binary data transfer */
    plcDatumOutput  outfunc;
    plcDatumInput   infunc;

    /* GPDB in- and out- functions to transform custom types to text and back */
    RegProcedure    output, input;

    /* Information used for type input/output operations */
    Oid             typeOid;
    Oid             typelem;
    bool            typbyval;
    int16           typlen;
    char            typalign;
    int32           typmod;

    /* UDT-specific information */
    bool            is_rowtype;
    bool            is_record;
    bool            attisdropped;
    Oid             typ_relid;
    TransactionId   typrel_xmin;
    ItemPointerData typrel_tid;
    char           *typeName;
};

typedef struct plcPgArrayPosition {
    plcTypeInfo    *type;
    bits8          *bitmap;
    int             bitmask;
} plcPgArrayPosition;

void fill_type_info(FunctionCallInfo fcinfo, Oid typeOid, plcTypeInfo *type);
void copy_type_info(plcType *type, plcTypeInfo *ptype);
void free_type_info(plcTypeInfo *type);
char *fill_type_value(Datum funcArg, plcTypeInfo *argType);

#endif /* PLC_TYPEIO_H */
