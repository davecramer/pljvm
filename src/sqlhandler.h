/*------------------------------------------------------------------------------
 *
 *
 * Copyright (c) 2016, Pivotal.
 *
 *------------------------------------------------------------------------------
 */


#ifndef PLC_SQLHANDLER_H
#define PLC_SQLHANDLER_H

#include "common/messages/messages.h"

plcMessage *handle_sql_message(plcMsgSQL *msg);

#endif /* PLC_SQLHANDLER_H */
