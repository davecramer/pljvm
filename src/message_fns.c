/*
Copyright 1994 The PL-J Project. All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer
   in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE PL-J PROJECT ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
THE PL-J PROJECT OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
   OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
   OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
   OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those of the authors and should not be
interpreted as representing official policies, either expressed or implied, of the PL-J Project.
*/


/**
 * file name:        plj-callmkr.c
 * description:        PL/pgJ call message creator routine. This file
 *                    was renamed from plpgj_call_maker.c
 *                    It is replaceable with pljava-way of declaring java
 *                    method calls. (read the readme!)
 * author:        Laszlo Hornyak
 */

/* Greenplum headers */
#include "postgres.h"
#include "executor/spi.h"
#include "access/htup_details.h"
#include "access/transam.h"
#include "catalog/pg_proc.h"
#include "catalog/pg_type.h"
#include "utils/array.h"
#include "utils/syscache.h"
#include "utils/builtins.h"

/* message and function definitions */
#include "common/comm_utils.h"
#include "common/messages/messages.h"
#include "message_fns.h"
#include "function_cache.h"
#include "plc_typeio.h"

static bool plc_procedure_valid(plcProcInfo *proc, HeapTuple procTup);
static bool plc_type_valid(plcTypeInfo *type);
static void fill_callreq_arguments(FunctionCallInfo fcinfo, plcProcInfo *pinfo, plcMsgCallreq *req);

plcProcInfo * get_proc_info(FunctionCallInfo fcinfo) {
    int           i, len;
    Datum        *argnames = NULL;
    bool         *argnulls = NULL;
    Datum         argnamesArray;
    Datum         srcdatum, namedatum;
    bool          isnull;
    Oid           procoid;
    Form_pg_proc  procTup;
    HeapTuple     procHeapTup,
                  textHeapTup = NULL;
    Form_pg_type  typeTup;
    plcProcInfo  *pinfo = NULL;

    procoid = fcinfo->flinfo->fn_oid;
    procHeapTup = SearchSysCache(PROCOID, procoid, 0, 0, 0);
    if (!HeapTupleIsValid(procHeapTup)) {
        elog(ERROR, "cannot find proc with oid %u", procoid);
    }

    pinfo   = function_cache_get(procoid);
    /*
     * All the catalog operations are done only if the cached function
     * information has changed in the catalog
     */
    if (!plc_procedure_valid(pinfo, procHeapTup)) {

        /*
         * Here we are using plc_top_alloc as the function structure should be
         * available across the function handler call
         *
         * Note: we free the procedure from within function_put_cache below
         */
        pinfo = plc_top_alloc(sizeof(plcProcInfo));
        if (pinfo == NULL) {
            elog(FATAL, "Cannot allocate memory for plcProcInfo structure");
        }
        /* Remember transactional information to allow caching */
        pinfo->funcOid = procoid;
        pinfo->fn_xmin = HeapTupleHeaderGetXmin(procHeapTup->t_data);
        pinfo->fn_tid  = procHeapTup->t_self;
        pinfo->retset  = fcinfo->flinfo->fn_retset;
        pinfo->hasChanged = 1;

        procTup = (Form_pg_proc)GETSTRUCT(procHeapTup);
        fill_type_info(fcinfo, procTup->prorettype, &pinfo->rettype);

        pinfo->nargs = procTup->pronargs;
        if (pinfo->nargs > 0) {
            // This is required to avoid the cycle from being removed by optimizer
            int volatile j;

            pinfo->argtypes = plc_top_alloc(pinfo->nargs * sizeof(plcTypeInfo));
            for (j = 0; j < pinfo->nargs; j++) {
                fill_type_info(fcinfo, procTup->proargtypes.values[j], &pinfo->argtypes[j]);
            }

            argnamesArray = SysCacheGetAttr(PROCOID, procHeapTup,
                                            Anum_pg_proc_proargnames, &isnull);
            /* If at least some arguments have names */
            if (!isnull) {
                textHeapTup = SearchSysCache(TYPEOID, ObjectIdGetDatum(TEXTOID), 0, 0, 0);
                if (!HeapTupleIsValid(textHeapTup)) {
                    elog(FATAL, "cannot find text type in cache");
                }
                typeTup = (Form_pg_type)GETSTRUCT(textHeapTup);
                deconstruct_array(DatumGetArrayTypeP(argnamesArray), TEXTOID,
                                  typeTup->typlen, typeTup->typbyval, typeTup->typalign,
                                  &argnames, &argnulls, &len);
                if (len != pinfo->nargs) {
                    elog(FATAL, "something bad happened, nargs != len");
                }
            }

            pinfo->argnames = plc_top_alloc(pinfo->nargs * sizeof(char*));
            for (i = 0; i < pinfo->nargs; i++) {
                if (!isnull && !argnulls[i]) {
                    pinfo->argnames[i] =
                        plc_top_strdup(DatumGetCString(
                                DirectFunctionCall1(textout, argnames[i])
                            ));
                    if (strlen(pinfo->argnames[i]) == 0) {
                        pfree(pinfo->argnames[i]);
                        pinfo->argnames[i] = NULL;
                    }
                } else {
                    pinfo->argnames[i] = NULL;
                }
            }

            if (textHeapTup != NULL) {
                ReleaseSysCache(textHeapTup);
            }
        } else {
            pinfo->argtypes = NULL;
            pinfo->argnames = NULL;
        }

        /* Get the text and name of the function */
        srcdatum = SysCacheGetAttr(PROCOID, procHeapTup, Anum_pg_proc_prosrc, &isnull);
        if (isnull)
            elog(ERROR, "null prosrc");
        pinfo->src = plc_top_strdup(DatumGetCString(DirectFunctionCall1(textout, srcdatum)));
        namedatum = SysCacheGetAttr(PROCOID, procHeapTup, Anum_pg_proc_proname, &isnull);
        if (isnull)
            elog(ERROR, "null proname");
        pinfo->name = plc_top_strdup(DatumGetCString(DirectFunctionCall1(nameout, namedatum)));

        /* Cache the function for later use */
        function_cache_put(pinfo);
    } else {
        pinfo->hasChanged = 0;
    }
    ReleaseSysCache(procHeapTup);
    return pinfo;
}

void free_proc_info(plcProcInfo *proc) {
    int i;
    for (i = 0; i < proc->nargs; i++) {
        if (proc->argnames[i] != NULL) {
            pfree(proc->argnames[i]);
        }
        free_type_info(&proc->argtypes[i]);
    }
    if (proc->nargs > 0) {
        pfree(proc->argnames);
        pfree(proc->argtypes);
    }
    pfree(proc);
}

plcMsgCallreq *pljvm_create_call(FunctionCallInfo fcinfo, plcProcInfo *pinfo) {
    plcMsgCallreq *req;

    req          = pmalloc(sizeof(plcMsgCallreq));
    req->msgtype = MT_CALLREQ;
    req->proc.name = pinfo->name;
    req->proc.src  = pinfo->src;
    req->objectid  = pinfo->funcOid;
    req->hasChanged = pinfo->hasChanged;
    copy_type_info(&req->retType, &pinfo->rettype);

    fill_callreq_arguments(fcinfo, pinfo, req);

    return req;
}

static bool plc_type_valid(plcTypeInfo *type) {
    bool valid = true;
    int  i;

    for (i = 0; i < type->nSubTypes && valid; i++) {
        valid = plc_type_valid(&type->subTypes[i]);
    }

    /* We exclude record from testing here, as it would change only if the function
     * changes itself, which would be caugth by checking function creation xid */
    if (valid && type->is_rowtype && !type->is_record) {
        HeapTuple relTup;

        Assert(OidIsValid(type->typ_relid));
        Assert(TransactionIdIsValid(type->typrel_xmin));
        Assert(ItemPointerIsValid(&type->typrel_tid));

        // Get the pg_class tuple for the argument type
        relTup = SearchSysCache1(RELOID, ObjectIdGetDatum(type->typ_relid));
        if (!HeapTupleIsValid(relTup))
            elog(ERROR, "PL/Container cache lookup failed for relation %u", type->typ_relid);

        // If commit transaction ID has changed or relation was moved within table
        // our type information is no longer valid
        if (type->typrel_xmin != HeapTupleHeaderGetXmin(relTup->t_data) ||
                !ItemPointerEquals(&type->typrel_tid, &relTup->t_self)) {
            valid = false;
        }

        ReleaseSysCache(relTup);
    }

    return valid;
}

/*
 * Decide whether a cached PLyProcedure struct is still valid
 */
static bool plc_procedure_valid(plcProcInfo *proc, HeapTuple procTup) {
    bool valid = false;

    if (proc != NULL) {
        /* If the pg_proc tuple has changed, it's not valid */
        if (proc->fn_xmin == HeapTupleHeaderGetXmin(procTup->t_data) &&
            ItemPointerEquals(&proc->fn_tid, &procTup->t_self)) {
            int  i;

            valid = true;

            // If there are composite input arguments, they might have changed
            for (i = 0; i < proc->nargs && valid; i++) {
                valid = plc_type_valid(&proc->argtypes[i]);
            }

            // Also check for composite output type
            if (valid) {
                valid = plc_type_valid(&proc->rettype);
            }
        }
    }
    return valid;
}

static void fill_callreq_arguments(FunctionCallInfo fcinfo, plcProcInfo *pinfo, plcMsgCallreq *req) {
    int   i;

    req->nargs = pinfo->nargs;
    req->retset = pinfo->retset;
    req->args  = pmalloc(sizeof(*req->args) * pinfo->nargs);

    for (i = 0; i < pinfo->nargs; i++) {
        req->args[i].name = pinfo->argnames[i];
        copy_type_info(&req->args[i].type, &pinfo->argtypes[i]);

        if (fcinfo->argnull[i]) {
            req->args[i].data.isnull = 1;
            req->args[i].data.value = NULL;
        } else {
            req->args[i].data.isnull = 0;
            req->args[i].data.value = pinfo->argtypes[i].outfunc(fcinfo->arg[i], &pinfo->argtypes[i]);
        }
    }
}
