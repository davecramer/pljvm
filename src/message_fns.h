/*
Copyright 1994 The PL-J Project. All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer
   in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE PL-J PROJECT ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
THE PL-J PROJECT OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
   OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
   OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
   OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those of the authors and should not be
interpreted as representing official policies, either expressed or implied, of the PL-J Project.
*/

#ifndef PLC_MESSAGE_FNS_H
#define PLC_MESSAGE_FNS_H

#include "postgres.h"
#include "fmgr.h"

#include "common/messages/messages.h"
#include "plc_typeio.h"

/* Structure representing function result data */
typedef struct {
    plcMsgResult    *resmsg;
    int              resrow;
} plcProcResult;

typedef struct {
    /* Greenplum Function Information */
    Oid              funcOid;
    TransactionId    fn_xmin; /* Transaction ID that created this function in catalog */
    ItemPointerData  fn_tid;  /* ItemPointer for the function row in catalog */
    /* Universal Function Information */
    char            *name;
    char            *src;
    int              hasChanged; /* Whether the function has changed since last call */
    plcTypeInfo      rettype;
    int              retset;
    int              nargs;
    char           **argnames;
    plcTypeInfo     *argtypes;
} plcProcInfo;

plcProcInfo *get_proc_info(FunctionCallInfo fcinfo);
void free_proc_info(plcProcInfo *proc);

plcMsgCallreq *pljvm_create_call(FunctionCallInfo fcinfo, plcProcInfo *pinfo);

#endif /* PLC_MESSAGE_FNS_H */
