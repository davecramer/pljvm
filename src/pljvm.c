/*------------------------------------------------------------------------------
 *
 *
 * Copyright (c) 2016, Pivotal.
 *
 *------------------------------------------------------------------------------
 */


/* Postgres Headers */
#include "postgres.h"
#include "fmgr.h"
#include "funcapi.h"
#include "miscadmin.h"
#include "executor/spi.h"
#include "commands/trigger.h"


/* PLjvm Headers */
#include "common/comm_channel.h"
#include "common/messages/messages.h"
#include "message_fns.h"
#include "sqlhandler.h"
#include "plc_typeio.h"
#include "pljvm.h"

#ifdef PG_MODULE_MAGIC
PG_MODULE_MAGIC;
#endif

static int port;
static char *host;

PG_FUNCTION_INFO_V1(pljvm_call_handler);

static Datum pljvm_call_hook(PG_FUNCTION_ARGS);
static plcProcResult *pljvm_get_result(FunctionCallInfo  fcinfo,
                                             plcProcInfo      *pinfo);
static Datum pljvm_process_result(FunctionCallInfo  fcinfo,
                                        plcProcInfo      *pinfo,
                                        plcProcResult    *presult);
static void pljvm_process_exception(plcMsgError *msg);
static void pljvm_process_sql(plcMsgSQL *msg, plcConn* conn);
static void pljvm_process_log(plcMsgLog *log);

Datum pljvm_call_handler(PG_FUNCTION_ARGS) {
    Datum datumreturn = (Datum) 0;
    MemoryContext oldMC = NULL;
    int ret;

    /* TODO: handle trigger requests as well */
    if (CALLED_AS_TRIGGER(fcinfo)) {
        elog(ERROR, "PL/Container does not support triggers");
        return datumreturn;
    }

    /* save caller's context */
    oldMC = pl_container_caller_context;
    pl_container_caller_context = CurrentMemoryContext;

    /* Create a new memory context and switch to it */
    ret = SPI_connect();
    if (ret != SPI_OK_CONNECT)
        elog(ERROR, "[pljvm] SPI connect error: %d (%s)", ret,
             SPI_result_code_string(ret));

    /* We need to cover this in try-catch block to catch the even of user
     * requesting the query termination. In this case we should forcefully
     * kill the container and reset its information
     */
    PG_TRY();
    {
        datumreturn = pljvm_call_hook(fcinfo);
    }
    PG_CATCH();
    {
        /* If the reason is Cancel or Termination */
        if (InterruptPending || QueryCancelPending
#ifdef GP_VERSION
        		|| QueryFinishPending
#endif
				) {
        }
        PG_RE_THROW();
    }
    PG_END_TRY();

    /* Return to old memory context */
    ret = SPI_finish();
    if (ret != SPI_OK_FINISH)
        elog(ERROR, "[pljvm] SPI finish error: %d (%s)", ret,
             SPI_result_code_string(ret));

    pl_container_caller_context = oldMC;
    return datumreturn;
}

static Datum pljvm_call_hook(PG_FUNCTION_ARGS) {
    Datum                     result = (Datum) 0;
    plcProcInfo              *pinfo;
    bool                      bFirstTimeCall = true;
    FuncCallContext *volatile funcctx = NULL;
    MemoryContext             oldcontext = NULL;
    plcProcResult            *presult = NULL;

    /* By default we return NULL */
    fcinfo->isnull = true;

    /* Get procedure info from cache or compose it based on catalog */
    pinfo = get_proc_info(fcinfo);

    /* If we have a set-returning function */
    if (fcinfo->flinfo->fn_retset) {
        /* First Call setup */
        if (SRF_IS_FIRSTCALL()) {
            funcctx = SRF_FIRSTCALL_INIT();
        } else {
            bFirstTimeCall = false;
        }

        /* Every call setup */
        funcctx = SRF_PERCALL_SETUP();
        Assert(funcctx != NULL);

        /* SRF initializes special context shared between function calls */
        oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
    } else {
        oldcontext = MemoryContextSwitchTo(pl_container_caller_context);
    }

    /* First time call for SRF or just a call of scalar function */
    if (bFirstTimeCall) {
        presult = pljvm_get_result(fcinfo, pinfo);
        if (fcinfo->flinfo->fn_retset) {
            funcctx->user_fctx = (void*)presult;
        }
    } else {
        presult = (plcProcResult*)funcctx->user_fctx;
    }

    if (presult == NULL) {
        MemoryContextSwitchTo(oldcontext);
        return result;
    }
    /* If we processed all the rows or the function returned 0 rows we can return immediately */
    if (presult->resrow >= presult->resmsg->rows) {
        free_result(presult->resmsg, false);
        pfree(presult);
        MemoryContextSwitchTo(oldcontext);
        SRF_RETURN_DONE(funcctx);
    }

    /* Process the result message from client */
    result = pljvm_process_result(fcinfo, pinfo, presult);

    presult->resrow += 1;
    MemoryContextSwitchTo(oldcontext);

    if (fcinfo->flinfo->fn_retset) {
        SRF_RETURN_NEXT(funcctx, result);
    } else {
        free_result(presult->resmsg, false);
        pfree(presult);
    }

    return result;
}

static plcProcResult *pljvm_get_result(FunctionCallInfo  fcinfo,
                                             plcProcInfo      *pinfo) {
    plcConn       *conn;
    int            message_type;
    plcMsgCallreq *req    = NULL;
    plcProcResult *result = NULL;

    conn = plcConnect(8000);

    if (conn != NULL) {
        req = pljvm_create_call(fcinfo, pinfo);
        pljvm_channel_send(conn, (plcMessage*)req);
        free_callreq(req, true, true);

        while (1) {
            int res = 0;
            plcMessage *answer;

            res = pljvm_channel_receive(conn, &answer);
            if (res < 0) {
                elog(ERROR, "Error receiving data from the client, %d", res);
                break;
            }

            message_type = answer->msgtype;
            switch (message_type) {
                case MT_RESULT:
                    result = (plcProcResult*)pmalloc(sizeof(plcProcResult));
                    result->resmsg = (plcMsgResult*)answer;
                    result->resrow = 0;
                    break;
                case MT_EXCEPTION:
                    pljvm_process_exception((plcMsgError*)answer);
                    break;
                case MT_SQL:
                    pljvm_process_sql((plcMsgSQL*)answer, conn);
                    break;
                case MT_LOG:
                    pljvm_process_log((plcMsgLog*)answer);
                    break;
                default:
                    elog(ERROR, "Received unhandled message with type id %d "
                    "from client", message_type);
                    break;
            }

            if (message_type != MT_SQL && message_type != MT_LOG)
                break;
        }
        plcDisconnect(conn);
    }
    return result;
}

/*
 * Processing client results message
 */
static Datum pljvm_process_result(FunctionCallInfo  fcinfo,
                                        plcProcInfo      *pinfo,
                                        plcProcResult    *presult) {
    Datum         result = (Datum) 0;
    plcMsgResult *resmsg = presult->resmsg;

    if (resmsg->cols > 1) {
        elog(ERROR, "Functions returning multiple columns are not supported yet");
        return result;
    }

    if (resmsg->rows == 0) {
        return result;
    }

    if (presult->resrow >= resmsg->rows) {
        ereport(ERROR,
             (errcode(ERRCODE_CARDINALITY_VIOLATION),
             errmsg( "Trying to access result row %d of the %d-rows result set",
                     presult->resrow, resmsg->rows)));
        return result;
    }

    if (resmsg->data[presult->resrow][0].isnull == 0) {
        fcinfo->isnull = false;
        result = pinfo->rettype.infunc(resmsg->data[presult->resrow][0].value, &pinfo->rettype);
    }

    return result;
}

/*
 * Processing client log message
 */
static void pljvm_process_log(plcMsgLog *log) {
    ereport(log->level,
            (errcode(ERRCODE_NO_DATA),
                    errmsg("%s", log->message)));
}

/*
 * Processing client SQL query message
 */
static void pljvm_process_sql(plcMsgSQL *msg, plcConn* conn) {
    plcMessage *res;
    volatile MemoryContext oldcontext;
    volatile ResourceOwner oldowner;

    oldcontext = CurrentMemoryContext;
    oldowner = CurrentResourceOwner;
    MemoryContextSwitchTo(pl_container_caller_context);

    res = handle_sql_message(msg);
    if (res != NULL) {
        pljvm_channel_send(conn, res);
        switch (res->msgtype) {
            case MT_RESULT:
                free_result((plcMsgResult*)res, true);
                break;
            case MT_CALLREQ:
                free_callreq((plcMsgCallreq*)res, true, true);
                break;
            default:
                ereport(ERROR,
                      (errcode(ERRCODE_RAISE_EXCEPTION),
                      errmsg( "Returning message type '%c' from SPI call is not implemented", res->msgtype)));
        }
    }

    MemoryContextSwitchTo(oldcontext);
    CurrentResourceOwner = oldowner;

    /*
     * AtEOSubXact_SPI() should not have popped any SPI context, but just
     * in case it did, make sure we remain connected.
     */
    SPI_restore_connection();
}

/*
 * Processing client exception message
 */
static void pljvm_process_exception(plcMsgError *msg) {
    if (msg->stacktrace != NULL) {
        ereport(ERROR,
                (errcode(ERRCODE_RAISE_EXCEPTION),
                 errmsg("PL/JVM client exception occurred: \n %s \n %s", msg->message, msg->stacktrace)));
    } else {
        ereport(ERROR,
                (errcode(ERRCODE_RAISE_EXCEPTION),
                 errmsg("PL/JVM client exception occurred: \n %s", msg->message)));
    }
    free_error(msg);
}
