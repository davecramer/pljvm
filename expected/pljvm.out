CREATE EXTENSION pljvm;
CREATE OR REPLACE FUNCTION pljvm_log100(d float8) RETURNS float8 AS $$
org.postgresql.plj.test.Math.log10
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_bool(b bool) RETURNS bool AS $$
org.postgresql.plj.test.Bool.valueOf
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_add(i int2, j int2) RETURNS int2 AS $$
org.postgresql.plj.test.Int.add
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_add(i int4, j int4) RETURNS int4 AS $$
org.postgresql.plj.test.Int.add
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_add(i int8, j int8) RETURNS int8 AS $$
org.postgresql.plj.test.Int.add
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_add(f float4, g float4) RETURNS float4 AS $$
org.postgresql.plj.test.Math.add
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_add(f float8, g float8) RETURNS float8 AS $$
org.postgresql.plj.test.Math.add
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_numeric(n1 numeric, n2 numeric) RETURNS numeric AS $$
org.postgresql.plj.test.Math.add
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_timestamp(t timestamp) RETURNS timestamp AS $$
org.postgresql.plj.test.Timestamp.valueOf
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_timestamptz(t timestamptz) RETURNS timestamptz AS $$
org.postgresql.plj.test.Timestamp.valueOf
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_text(arg1 varchar, arg2 varchar) RETURNS varchar AS $$
org.postgresql.plj.test.Text.concat
$$ LANGUAGE pljvm;
CREATE OR REPLACE FUNCTION pljvm_add_arr(arg1 int4[], int4) returns int4[] as $$
org.postgresql.plj.test.Int.add
$$ language 'pljvm';
CREATE OR REPLACE FUNCTION pljvm_add_arr(arg1 int2[], int2) returns int2[] as $$
org.postgresql.plj.test.Int.add
$$ language 'pljvm';
CREATE OR REPLACE FUNCTION pljvm_add_arr(arg1 int8[], int4) returns int8[] as $$
org.postgresql.plj.test.Int.add
$$ language 'pljvm';
CREATE OR REPLACE FUNCTION pljvm_bool_arr_xor(arg1 boolean[], x boolean) returns boolean[] as $$
org.postgresql.plj.test.Bool.xor
$$ language 'pljvm';
CREATE OR REPLACE FUNCTION pljvm_exception() returns void as $$
org.postgresql.plj.test.Error.nullPointer
$$ language 'pljvm';
CREATE OR REPLACE FUNCTION pljvm_log(msg text) returns void as $$
org.postgresql.plj.test.Log.log
$$ language 'pljvm';
select pljvm_log100(100);
 pljvm_log100 
--------------
            2
(1 row)

select pljvm_bool(true);
 pljvm_bool 
------------
 t
(1 row)

select pljvm_add(1::int2,2::int2);
 pljvm_add 
-----------
         3
(1 row)

select pljvm_add(3,4);
 pljvm_add 
-----------
         7
(1 row)

select pljvm_add(5::int8,6::int8);
 pljvm_add 
-----------
        11
(1 row)

select pljvm_add(1.2::float4,3.4::float4);
 pljvm_add 
-----------
       4.6
(1 row)

select pljvm_add(5.6::float8,7.8::float8);
 pljvm_add 
-----------
      13.4
(1 row)

select pljvm_timestamp('2017-11-24 07:23:20.222432');
         pljvm_timestamp         
---------------------------------
 Fri Nov 24 07:23:20.222432 2017
(1 row)

select pljvm_timestamptz('2017-11-24 07:23:48.597243-08');
          pljvm_timestamptz          
-------------------------------------
 Fri Nov 24 07:23:48.597243 2017 PST
(1 row)

select pljvm_text('hello ', 'world');
 pljvm_text  
-------------
 hello world
(1 row)

select pljvm_add_arr('{1,2,3}'::int2[], 1::int2);
 pljvm_add_arr 
---------------
 {2,3,4}
(1 row)

select pljvm_add_arr('{3,4,5}'::int4[], 1);
 pljvm_add_arr 
---------------
 {4,5,6}
(1 row)

select pljvm_add_arr('{6,7,8}'::int8[], 1);
 pljvm_add_arr 
---------------
 {7,8,9}
(1 row)

select pljvm_bool_arr_xor('{1,0,0,1}'::boolean[],true);
 pljvm_bool_arr_xor 
--------------------
 {f,t,t,f}
(1 row)

select pljvm_log('hello');
INFO:  org.postgresql.plj.test.Log-log-hello
 pljvm_log 
-----------
 
(1 row)

